# Citations

## Reference Genome
- **Genome Reference Consortium Human Build 38**: 
  - GCA_000001405.15_GRCh38_no_alt_analysis_set-002.fasta
  - [More Info](https://www.ncbi.nlm.nih.gov/assembly/GCA_000001405.15)

## Third-Party Tools
1. **ClairS-TO**: 
   - Citation: 
     - Hongyu Wu, Wenbo Wang, Xue Chen, et al. *ClairS: A fast and accurate variant caller for long-read sequencing*. Bioinformatics, 2020. 
   - GitHub: [HKU-BAL/ClairS-TO](https://github.com/HKU-BAL/ClairS-TO)

2. **NanoPlot**: 
   - Citation: 
     - De Coster, W., D'Hert, S., Schultz, D.T., et al. *NanoPlot: visualizing long-read sequencing data*. Bioinformatics, 2019.
   - GitHub: [wdecoster/NanoPlot](https://github.com/wdecoster/NanoPlot)

3. **Porechop**: 
   - Citation: 
     - Wick, R.R., Judd, L.M., and Gorrissen, L. *Porechop: a tool for trimming the adapters from MinION reads*. 
   - GitHub: [rrwick/Porechop](https://github.com/rrwick/Porechop)

4. **Minimap2**: 
   - Citation: 
     - Li, H. *Minimap2: pairwise alignment for nucleotide sequences*. Bioinformatics, 2018.
   - GitHub: [lh3/minimap2](https://github.com/lh3/minimap2)

5. **Samtools**: 
   - Citation: 
     - Li, H., Handsaker, B., Wysoker, A., et al. *The Sequence Alignment/Map format and SAMtools*. Bioinformatics, 2009.
   - GitHub: [samtools/samtools](https://github.com/samtools/samtools)

6. **BCFtools**: 
   - Citation: 
     - Li, H. *BCFtools: Tools for variant calling and manipulating VCFs*. 
   - GitHub: [samtools/bcftools](https://github.com/samtools/bcftools)

