# SomaTar
SomaTar is tailored for the detection of somatic variants in tumor only samples using long-read sequencing data from Oxford Nanopore technology, employing ClairS-TO for variant calling to enhance accuracy and resolution in identifying complex genomic alterations. 

# Getting Started
**Prerequisites** :  

Docker  
Python  
R  

  **Variant Caller**  
-  ClairS-TO (https://github.com/HKU-BAL/ClairS-TO)

   ClairS-TO (Somatic Tumor-Only) is a tool in the Clair series to support long-read somatic small variant calling with only tumor samples available.
   
   **Installation** :  
   docker pull hkubal/clairs-to

**Other tools needed to be pre-installed**

- Nanoplot (https://github.com/wdecoster/NanoPlot)
- Porechop (https://github.com/rrwick/Porechop)  
- Minimap2 (https://github.com/lh3/minimap2)
- Samtools (https://github.com/samtools/samtools) 
- BCFtools (https://samtools.github.io/bcftools/howtos/index.html)

# Overview of the Pipeline  

The SomaTar pipeline consists of three scripts.:  
1) gunzip.sh
2) merge_fastq.sh
3) SomaTar.sh
     
# 1) gunzip.sh : 

The fastq files generated by sequencer are zipped files. They need to be unzipped. The gunzip.sh script can be used to perform unzipping of fastq files.  

```bash 
Usage :  ./gunzip.sh  [-w working dir] 
Where,  
-w is the working directory where all barcode folders are located after sequencing 
 ```      

# 2) merge_fastq.sh : 

Being ONT a real time sequencing technology, it generates multiple fastq files for the same barcode. All fastq files need to be concatenated into one fastq file. Also, the script has a utility to supply a matadata.csv file to rename barcodes with their sample ids.  
```bash
Usage :  ./cat.sh  [-w working dir] [-d destination dir for merged fastq]  [-m matadata file]**  
Where,  
-w is the working directory where all barcode folders are located  
-d is a destination directory where all the merged fastq samples needs to go  
-m metadata file is metadata.csv file which contains the barcode names to be replaced with sample names.  

  Example of matadata.csv file:  
    barcode01	sample01  
    barcode02	sample02
```
# 3) SomaTar.sh : 
   This is the main script which contains the whole pipeline workflow. Above both scripts can be skipped if the fastq files are already merged and names of the barcode to samples can be changed manually.

```bash
Usage: ./SomaTar.sh [-w working dir] [-r reference] [-b bed file] [-t threads] [-q genotype quality] [-p platform]

Where:
-w specifies the working directory where all the merged fastq files are located
-r is for reference genome file (GCA_000001405.15_GRCh38_no_alt_analysis_set-002.fasta with its index file. 
   It has to be in the Clairs-To variant caller directory, generated as "ClairS_TO" after installation)
-b BED file for region of interest
-t threads represent maximum threads to be used
-q quality represents the genotype quality, suggested quality for ONT platform is 12, but it can be set as per the need
-p The platform can be set based on the ONT chemistry, instruments and basecaller being used during the sequencing. They can be selected from:
- ont_r10_dorado_sup_5khz_ssrs
- ont_r10_dorado_sup_5khz_ss
- ont_r10_dorado_sup_5khz
- ont_r10_dorado_sup_4khz
- ont_r10_dorado_hac_4khz
- ont_r10_guppy_sup_4khz
- ont_r10_guppy_hac_5khz

Porechop is optional. It can be enabled or disbaled with following options:

Enable porechopping :
 
Usage:./SomaTar.sh [-w working dir] [-r reference] [-b bed file] [-t threads] [-q genotype quality] [-p platform] [-pore_chop yes]

Disable porechopping :

Usage:./SomaTar.sh [-w working dir] [-r reference] [-b bed file] [-t threads] [-q genotype quality] [-p platform] [-pore_chop no] 

In case of enabling indel calling, no specification is needed as indel calling is default option in ClairS-To variant caller.

Indel calling with ClairS-TO is also optional. It is enabled by-default.

It can be disabled with following option:

./SomaTar.sh [-w working dir] [-r reference] [-b bed file] [-t threads] [-q genotype quality] [-p platform] [-pore_chop yes] [-disable_indel_calling]

```
**Please refer for more detail- https://github.com/HKU-BAL/ClairS-TO?tab=readme-ov-file#pre-trained-models**
  
**The key features of SomaTar.sh script are :**
```bash
1) Input Parameters:The script accepts several command-line options for configuration
    
2) Docker Permissions Check:The script checks Docker accessibility and prompts for sudo credentials if permissions on the Docker socket need to be fixed   

3) Quality Control with NanoPlot:It identifies FASTQ files in the specified directory and generates quality plots using NanoPlot for initial data assessment 

4) Trimming with Porechop: Barcodes are removed from the FASTQ files to ensure clean data for alignment    

5) Alignment with Minimap2:The trimmed FASTQ files are aligned to the reference genome, producing SAM files for further processing   

6) Sorting and Indexing:SAM files are converted to sorted BAM format using samtools, followed by indexing to facilitate efficient data retrieval   

7) Coverage and Depth Analysis:The script computes coverage statistics for specified regions based on the BED file and generates depth information for each BAM file  

8) SNP Calling with ClairS-TO: Utilizing Docker, the ClairS-TO pipeline is run for variant calling on each sorted BAM file, enabling comprehensive SNP detection    

9)  VCF to CSV Conversion: The resulting VCF files from ClairS-TO are converted into a user-friendly CSV format for easier analysis    

10) SNP Plotting: It plots VAF and RAF against their positions where SNPs are detected, offering a clear visual representation of allele frequencies across the genome to analyze the trand for potential LOH    
```

